Move kubeadm integration tests to test/integration/kubeadm

brianpursley opened the pull request:
Refactor to make kubeadm integration tests follow standard integration test pattern
- Moved kubeadm tests to test/integration/kubeadm directory
- Removed special case handling of kubeadm from hack/make-rules/test-cmd.sh
- Remove cmd/kubeadm/test/** exclusion from hack/make-rules/test.sh
- Build kubeadm and set KUBEADM_PATH in hack/make-rules/test-integration.sh so kubeadm integration tests can run
- Separated unit-test-specific util functions from integration-test-specific util functions

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
kubeadm tests in cmd/kubeadm/test are not true integration tests, and this causes a couple of issues.
1. Makes it difficult to use `go test ./...` in some cases because it tries to run these tests
2. Special code exists for kubeadm to exclude or include these tests for `make test` and `make test-integration`

**Which issue(s) this PR fixes**:
Fixes #91740 

**Special notes for your reviewer**:
This improves the usage of `go test ./...`, and fixes the problem as it pertains to kubeadm. **But using `go test ./...` from the project root *still* does not work because it will try to run the tests contained within test/e2e and test/integration**... it just tries to run everything it can find.

However, I think this is still a worthwhile refactor as it eliminates the need to handle kubeadm differently than every other part of the system when it comes to running unit and integration tests.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs

```

brianpursley commented:
/retest

brianpursley commented:
/retest

brianpursley commented:
/retest

neolit123 commented on hack/make-rules/test-cmd.sh:
```diff
@@ -116,28 +116,6 @@ function create_node() {
-    exit 0
-  fi
-fi
-
```
noting that this change needs an update in the k/kubeadm docs:
https://github.com/kubernetes/kubeadm/blob/c4674ed58e9116933e7a6d9a613c7a3d3ba4da8e/CONTRIBUTING.md#running-unit-tests

brianpursley replied:
Hmm, I wonder if I did this correctly...

Maybe it should still be run the same way, just have the code be located in the new directory.

This really is a strange test compared to the others because it requires the kubeadm executable to be built and located.

I'm not so sure now... :/

neolit123 replied:
leaving it under `make test-cmd` seems fine.

brianpursley replied:
If test-cmd is "command line integration tests", it feels like it still belongs there.

I don't know where the right place for the code is... test/integration directory seems like it has tests for things that are not command line.

test/cmd sort of seems like where it belongs, but that directory only has shell scripts.

brianpursley replied:
OK, let me push an update, keep it in test-cmd so not cause any usage changes

neolit123 replied:
having a folder called kubeadm under `./test/cmd` seems OK to me.
leaving this decision to SIG Testing and the maintainers of /test

the alternative solution as mentioned by liggitt here:
https://github.com/kubernetes/kubernetes/issues/91740#issuecomment-638993147

is to just skip the kubeadm integration tests if the env. variable KUBEADM_PATH is not set.
and leave the kubeadm tests under cmd/kubeadm/test

neolit123 commented on hack/make-rules/test-integration.sh:
```diff
@@ -66,6 +66,15 @@ runTests() {
+    make all -C "${KUBE_ROOT}" WHAT=cmd/kubeadm
+    KUBEADM_PATH="$(kube::realpath "${KUBE_ROOT}")/_output/local/go/bin/kubeadm"
+    export KUBEADM_PATH
+  fi
```
[1]

neolit123 commented on test/integration/kubeadm/main_test.go:
```diff
@@ -0,0 +1,27 @@
+)
+
+func TestMain(m *testing.M) {
+	framework.EtcdMain(m.Run)
```
the kubeadm integration tests are CLI integration tests, that don't create a real cluster.
therefore setup of etcd is not required.

neolit123 commented:
>  ...stdata/init/current-component-config.yaml → ...stdata/init/current-component-config.yaml 

we shouldn't move kubeadm API specifics in test/integration
https://github.com/kubernetes/kubernetes/blob/e911bf7bfae5407402ecd1bd6eafd4f017dd8dec/test/integration/kubeadm/testdata/init/current-component-config.yaml

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -30,9 +30,9 @@ import (
 	certsphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/certs"
 	kubeconfigphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig"
+	testutil "k8s.io/kubernetes/cmd/kubeadm/app/testing"
+	cmdtestutil "k8s.io/kubernetes/cmd/kubeadm/app/testing/cmd"
```
so the introduction of cmd/kubeadm/app/testing is something good and i forgot about this problem where kubeadm imports it's integration tests for some utils.
i wasn't around to gate such a change, so it slipped trough in multiple packages...

so by moving such utils under /app, /app no longer has to import /test which is good.

neolit123 commented:
/cc @rosti @fabriziopandini

brianpursley commented:
I just found one more special-case for kubeadm that can be removed...
https://github.com/kubernetes/kubernetes/blob/a36b9a992223a71d822f3a0680f6b5202d2687d8/test/cmd/legacy-script.sh#L422

brianpursley commented:
Thanks for all the comments so far. I think I didn't fully "get" the purpose and difference between test-cmd and test-integration and how it pertains to kubeadm, but I think I understand a little better now.  I will make some changes to keep usage the same...

brianpursley committed:
Refactor to move kubeadm integration tests to test/cmd/kubeadm
- Moved kubeadm tests to test/cmd/kubeadm directory
- Remove cmd/kubeadm/test/** exclusion from hack/make-rules/test.sh
- Separated unit-test-specific util functions from integration-test-specific util functions

brianpursley commented:
/retest

brianpursley commented:
I have some doubts about this PR... Definitely more than I intended.

/Hold
